lib/block_view_validator.go

```diff
@@ -0,0 +1,1415 @@
+	if err != nil {
+		return 0, 0, nil, errors.Wrapf(err, "_connectUnregisterAsValidator: ")
+	}
+	// Note that we don't need to check isDeleted because the Get returns nil if isDeleted=true.
+	if prevValidatorEntry == nil {
+		return 0, 0, nil, RuleErrorValidatorNotFound
+	}
+	bav._deleteValidatorEntryMappings(prevValidatorEntry)
+
+	// Add a UTXO operation.
+	utxoOpsForTxn = append(utxoOpsForTxn, &UtxoOperation{
+		Type:               OperationTypeUnregisterAsValidator,
+		PrevValidatorEntry: prevValidatorEntry,
+		// PrevStakeEntries: prevStakeEntries, // TODO: in subsequent PR
+	})
+	return totalInput, totalOutput, utxoOpsForTxn, nil
+}
+
+func (bav *UtxoView) _disconnectUnregisterAsValidator(
+	operationType OperationType,
+	currentTxn *MsgDeSoTxn,
+	txHash *BlockHash,
+	utxoOpsForTxn []*UtxoOperation,
+	blockHeight uint32,
+) error {
+	// Validate the last operation is an UnregisterAsValidator operation.
+	if len(utxoOpsForTxn) == 0 {
+		return fmt.Errorf("_disconnectUnregisterAsValidator: utxoOperations are missing")
+	}
+	operationIndex := len(utxoOpsForTxn) - 1
+	operationData := utxoOpsForTxn[operationIndex]
+	if operationData.Type != OperationTypeUnregisterAsValidator {
+		return fmt.Errorf(
+			"_disconnectUnregisterAsValidator: trying to revert %v but found %v",
+			OperationTypeUnregisterAsValidator,
+			operationData.Type,
+		)
+	}
+
+	// Restore the PrevValidatorEntry.
+	prevValidatorEntry := operationData.PrevValidatorEntry
+	if prevValidatorEntry == nil {
+		// This should never happen as you can only unregister an existing ValidatorEntry
+		// when connecting. So disconnecting should always have a PrevValidatorEntry.
+		return fmt.Errorf(
+			"_disconnectUnregisterAsValidator: no deleted ValidatorEntry found for %v", currentTxn.PublicKey,
+		)
+	}
+	bav._setValidatorEntryMappings(prevValidatorEntry)
+
+	// TODO: In subsequent PR, restore the prev StakeEntries, if any.
+
+	// Disconnect the BasicTransfer.
+	return bav._disconnectBasicTransfer(
+		currentTxn, txHash, utxoOpsForTxn[:operationIndex], blockHeight,
+	)
+}
+
+func (bav *UtxoView) IsValidRegisterAsValidatorMetadata(transactorPublicKey []byte, metadata *RegisterAsValidatorMetadata) error {
+	// Validate ValidatorPKID.
+	transactorPKIDEntry := bav.GetPKIDForPublicKey(transactorPublicKey)
+	if transactorPKIDEntry == nil || transactorPKIDEntry.isDeleted {
+		return RuleErrorInvalidValidatorPKID
+	}
+
+	// Validate Domains.
+	if len(metadata.Domains) < 1 {
+		return RuleErrorValidatorNoDomains
+	}
+	if len(metadata.Domains) > MaxValidatorNumDomains {
+		return RuleErrorValidatorTooManyDomains
+	}
+	var domainStrings []string
+	for _, domain := range metadata.Domains {
+		_, err := url.ParseRequestURI(string(domain))
+		if err != nil {
+			return fmt.Errorf("%s: %v", RuleErrorValidatorInvalidDomain, domain)
+		}
+		domainStrings = append(domainStrings, string(domain))
+	}
+	if len(NewSet(domainStrings).ToSlice()) != len(domainStrings) {
```

Comment:
It's uniqueness of the entire domain. IMO this isn't a big deal since the validator controls this and it is in their interest to provide valid domains. And we cap the number of domains to 12. So if the validator were to include https://example.com, https://EXAMPLE.com, http://example.com, etc. They can do it. It shouldn't impact consensus as long as those are all valid domains.